Refactor to better support multiple pages.

halfnibble opened the pull request:
## Add Render Flexibility

The purpose of these changes are to make it easier to take advantage of wkhtmltopdf's ability to render multiple pages as a single PDF document.

- I have changed convert_to_pdf to accept a list instead of a string.
- Factored out some repeating logic in render_pdf_from_template.

johnraz commented on wkhtmltopdf/utils.py:
```diff
@@ -109,7 +109,7 @@ def wkhtmltopdf(pages, output=None, **kwargs):
     return check_output(ck_args, **ck_kwargs)
 
-def convert_to_pdf(filename, header_filename=None, footer_filename=None, cmd_options=None):
+def convert_to_pdf(filename_list, header_filename=None, footer_filename=None, cmd_options=None):
```
Maybe we could try to not break the function's backward compatibility by coercing the value to a list if not a list.

halfnibble replied:
Excellent idea.

I was debating with myself if I should do that, and whether or not I should keep the name `filename` or change it to `filename_list`. Anyone who calls convert_to_pdf directly could have broken compatibility. 

But then I was reading about the importance of naming, and thought to change the name to `filename_list`.

Either way works for me. Should I do both of the following?
1. Change the name back to 'filename'
2. Add the list coercion.

johnraz replied:
I feel exactly the same as you, can't make my mind about what's better: keep the naming consistent or keep the api unbroken... @maxpeterson any thoughts?

maxpeterson replied:
Can we do both, change the name back to 'filename' and add the list coercing as this will have best backwards compatibility.

maxpeterson replied:
It is probably worth adding a comment explaining that the argument should be a list.

johnraz replied:
Seems reasonable.

johnraz commented on wkhtmltopdf/utils.py:
```diff
@@ -119,52 +119,65 @@ def convert_to_pdf(filename, header_filename=None, footer_filename=None, cmd_opt
-        for f in filter(None, (input_file, header_file, footer_file)):
-            f.close()
+        if temporary_file is not None:
+            temporary_file.close()
```
:+1:

maxpeterson replied:
This looks like it will close the `temporary_file` before the function returns. Since `delete=(not debug)` this will result in the temporary file being deleted before it is used if `debug` is `False` (e.g. in production). 

Not sure if there is a test for this but we should probably be ensuring the tests are run with `DEBUG` off.

maxpeterson replied:
Maybe you could use `with` as explained in http://effbot.org/zone/python-with-statement.htm :

```
    class create_rendered_file:
        def __enter__(self):
            set things up
            return thing
        def __exit__(self, type, value, traceback):
            tear things down

    with create_rendered_file() as thing:
         some code
```

maxpeterson replied:
.... or did I miss something, it is kind of late?

johnraz replied:
I proposed to return the file (see the PR's main discussion), the with statement is elegant but stacking 3 of them would look dirty too imo

halfnibble replied:
@maxpeterson No, you are correct. I noticed the issue last night when I deployed it to my production website. I will either use "with" or follow @johnraz recommendation, make the other changes, and send an updated PR.

maxpeterson replied:
Should have read the full conversation. I like the DRYness and I think `with` can handle multiple `open`s but I don't think you can do them conditionally and so the nesting will be ugly  - I am happy with either approach.

```
with open('a', 'w') as a, open('b', 'w') as b:
    do_something()
```

johnraz commented:
Very nice change, I would have made a new method `render_pdf_from_templates` or something to go along with this change. What do you think?

halfnibble commented:
I have a custom method that I am now using with this code to create multiple pages. Basically, instead of a single `context` object, I pass a list of context objects. However, the use case could vary widely, so I'm not sure how useful my method would be.

For example, while I reuse the same template for all pages, someone else could use different templates. I also had to subclass PDFTemplateResponse and write my own version of PDFTemplateView.

johnraz commented:
I was more thinking of having something like 
```
render_pdf_from_templates(pages, header, footer, request, cmd_options)
```
where `pages` would be a list of 2 tuples like `[(template, context), (template,context),...]`.
The same principle could be applied to `header` and `footer` params (they could take a two tuple as value).

But if we go this way we should change `render_pdf_from_template` too to apply the 2 tuple thing.

Do you believe this would cover more use cases with this form?

If we decide to break the `convert_to_pdf` signature I think we can consider this, otherwise it should be part of another PR.

halfnibble commented:
I like the idea of having a list of one or more tuples containing template and context. And I like the name `pages`.

That change would only take a few minutes for me to implement. (Although I'm turning in for the night).

It might also inspire me to make another PR with a generic view for rendering multiple pages.

johnraz commented:
Sounds great! No rush though ;-)

halfnibble commented:
Woah! Do **_not_** accept this pull request.

I just discovered (in production) that the way I've refactored closing files causes the temp file to be deleted before its rendered (for DEBUG = False).

I will rethink this. I will also look into some way to automate the tests against DEBUG = False.

johnraz commented:
Ah yes indeed, you are closing the file (which indeed has the side effect of deleting them) before `convert_to_pdf` is called.

Returning the file instead of the name from `create_rendered_file` and closing everything as it was done before should be enough.

Again, no hurry for me ;-)

halfnibble commented:
I'm not sure I like where I put the class, and I'm not sure how you feel about using classes to handle file delete, but please consider these changes.

Instead of having a `create_rendered_file` method, I have a `RenderedFile` class. The `__del__` method should handle file deletions on `DEBUG=False`. So far my testing has confirmed this.

Also, reverted naming back to `filename` for backwards-compatibility. And I don't believe list coercion is necessary because the `wkhtmltopdf` method already does it.

Finally, I updated the test run file. I don't know if this is a good way to test multiple settings values. So please advise me of better options.

johnraz commented on wkhtmltopdf/utils.py:
```diff
@@ -113,58 +113,81 @@ def convert_to_pdf(filename, header_filename=None, footer_filename=None, cmd_opt
+            self.filename = self.temporary_file.name
+        except:
+            # In case something fails, return an empty filename string.
+            self.filename = ''
```
No need to catch the exception, if it fails we want to know it.

johnraz replied:
And the all pdf creation should fail ;-)

halfnibble replied:
Good point. Will remove try/except in this next commit.

johnraz commented on wkhtmltopdf/utils.py:
```diff
@@ -113,58 +113,81 @@ def convert_to_pdf(filename, header_filename=None, footer_filename=None, cmd_opt
-        # Clean up temporary files
-        for f in filter(None, (input_file, header_file, footer_file)):
-            f.close()
+    return convert_to_pdf(filename=[input_file.filename],
```
I guess you can drop the list here.

halfnibble replied:
True, the list **_is_** optional. I just wanted to start showing that list arguments are possible. And that the final argument sent to wkhtmltopdf will be a list. 

But functionally, it will work either way. I will remove the listing (listering?) here if you like.

halfnibble replied:
Perhaps it doesn't make sense to pass a list, since it is always going to be a single file passed in this function.

johnraz commented on wkhtmltopdf/tests/run.py:
```diff
@@ -32,19 +52,11 @@
-if failures:
-    sys.exit(1)
+# Run tests with False debug settings to test temporary file delete operations.
+run_tests()
```
I would run the tests with debug set to false only and use `override_settings` decorator where I want something specifically different.

halfnibble replied:
Why didn't I think of that? I will revert to the original run.py code, and simply change the debug settings to False.

johnraz commented:
I like the object solution! See my inline comments!

maxpeterson commented on wkhtmltopdf/utils.py:
```diff
@@ -113,58 +113,77 @@ def convert_to_pdf(filename, header_filename=None, footer_filename=None, cmd_opt
+    def __del__(self):
+        # Always close the temporary_file on object destruction.
+        if self.temporary_file is not None:
+            self.temporary_file.close()
```
It would be nice to have a test to assert that this happen.

maxpeterson replied:
Nice solution!

johnraz replied:
:+1: on adding a test for this specific case.

maxpeterson commented:
Looks good to merge!

halfnibble commented:
Ok, great. Let me know your decision on `filename=[input_file.filename]` as a list or not.

johnraz commented:
I would remove the list. We should document the usage of internal methods and add some usage example. Maybe in another issue.

johnraz commented:
sorry I put my comments directly on the commit ^^

Other than those last comments it does LGTM

halfnibble commented:
@johnraz I agree about adding a method to RenderedFile. In fact, I think a lot of the functionality in utils.py could be refactored to make better use of OOP. For example, the RenderedFile class should contain the method for rendering the temporary file.

However, the changes I have in mind would likely change the API, and I'm not sure how easy it would be to add an interface to keep the legacy API the same (or if that's even desirable).

So, I'm testing the waters with this small change as is. If you want me to explore more extensive changes, then I'm happy to do so. I use this library A LOT.

johnraz commented:
@halfnibble: This PR is already covering quite a bunch of changes and is doing it well.
Let's merge this and start another PR for any further refactoring.

johnraz commented:
oh, and huge :+1:  for the contrib :-)
